app/src/org/commcare/android/util/MarkupUtil.java

```diff
@@ -1,19 +1,180 @@
-    
+
+    /*
+     * Developer Preference helper classes
+     */
+
+    public static Spannable styleSpannable(Context c, String message){
+
+        if(DeveloperPreferences.isCssEnabled()){
+
+            return htmlspanner.fromHtml(MarkupUtil.getStyleString() + message);
+        }
+
+        return new SpannableString(generateMarkdown(c, message));
+
+    }
+
+    public static Spannable localizeStyleSpannable(Context c, String localizationKey){
+
+        if(DeveloperPreferences.isCssEnabled()){
+            return htmlspanner.fromHtml(MarkupUtil.getStyleString() + new String(Localization.get(localizationKey)));
+        }
+
+        return new SpannableString(localizeGenerateMarkdown(c, localizationKey));
+    }
+
+    public static Spannable localizeStyleSpannable(Context c, String localizationKey, String localizationArg){
+
+        if(DeveloperPreferences.isCssEnabled()){
+            return htmlspanner.fromHtml(MarkupUtil.getStyleString() + Localization.get(localizationKey, localizationArg));
+        }
+
+        return new SpannableString(localizeGenerateMarkdown(c, localizationKey, new String[] {localizationArg}));
+    }
+
+
+
+    public static Spannable localizeStyleSpannable(Context c, String localizationKey, String[] localizationArgs){
+        if(DeveloperPreferences.isCssEnabled()){
+            return htmlspanner.fromHtml(MarkupUtil.getStyleString() + Localization.get(localizationKey, localizationArgs));
+        }
+
+        return new SpannableString(localizeGenerateMarkdown(c, localizationKey, localizationArgs));
+    }
+
+    /*
+     * Markdown styling utils from Bypass
+     * https://github.com/Uncodin/bypass
+     */
+
+    public static CharSequence localizeGenerateMarkdown(Context c, String localizationKey){
+        CharSequence mSequence = generateMarkdown(c, new String(Localization.get(localizationKey)));
+        return mSequence;
+    }
+
+    public static CharSequence localizeGenerateMarkdown(Context c, String localizationKey, String[] localizationArgs){
+        CharSequence mSequence = generateMarkdown(c, Localization.get(localizationKey, localizationArgs));
+        return mSequence;
+    }
+
+    public static CharSequence generateMarkdown(Context c, String message){
+        Bypass bypass = new Bypass(c);
+        CharSequence mSequence = bypass.markdownToSpannable(convertCharacterEncodings(message));
+
+        System.out.println("2915 msequence: " + mSequence);
+
+        System.out.print("2915 msequence: " + mSequence);
+
+        return MarkupUtil.trimTrailingWhitespace(mSequence);
+    }
+
+    /*
+     * Perform conversions from encodings
+     */
+
+    public static String convertCharacterEncodings(String input){
+        return convertNewlines(convertPoundSigns(input));
+    }
+
+    /*
+     * Convert to newline
+     */
+
+    public static String convertNewlines(String input){
+
+        return input.replace("%n", System.getProperty("line.separator"));
```

Comment:
hey,

I'm not super comfortable with "%n" and "%p" as delimeters (It's totally reasonable to imagine that people might have "%n" in their text). Are those directly in markdown? If not, we should talk about how to do this differently